things/clients/redis/events.go

```diff
@@ -17,75 +29,289 @@ var (
 	val := map[string]interface{}{
-		"id":        cte.id,
-		"owner":     cte.owner,
-		"operation": thingCreate,
+		"operation":  clientCreate,
+		"id":         cce.ID,
+		"status":     cce.Status.String(),
+		"created_at": cce.CreatedAt,
 	}
 
-	if cte.name != "" {
-		val["name"] = cte.name
+	if cce.Name != "" {
+		val["name"] = cce.Name
 	}
-
-	if cte.metadata != nil {
-		metadata, err := json.Marshal(cte.metadata)
+	if len(cce.Tags) > 0 {
+		tags := fmt.Sprintf("[%s]", strings.Join(cce.Tags, ","))
+		val["tags"] = tags
+	}
+	if cce.Owner != "" {
+		val["owner"] = cce.Owner
+	}
+	if cce.Metadata != nil {
+		metadata, err := json.Marshal(cce.Metadata)
 		if err != nil {
 			return val
 		}
 
-		val["metadata"] = string(metadata)
+		val["metadata"] = metadata
+	}
+	if cce.Credentials.Identity != "" {
+		val["identity"] = cce.Credentials.Identity
 	}
 
 	return val
 }
 
 type updateClientEvent struct {
-	id       string
-	name     string
-	owner    string
-	tags     []string
-	metadata map[string]interface{}
+	mfclients.Client
+	operation string
 }
 
-func (ute updateClientEvent) Encode() map[string]interface{} {
+func (uce updateClientEvent) Encode() map[string]interface{} {
 	val := map[string]interface{}{
-		"id":        ute.id,
-		"operation": thingUpdate,
+		"operation":  clientUpdate + "." + uce.operation,
+		"updated_at": uce.UpdatedAt,
+		"updated_by": uce.UpdatedBy,
 	}
 
-	if ute.name != "" {
-		val["name"] = ute.name
+	if uce.ID != "" {
+		val["id"] = uce.ID
 	}
-
-	if ute.metadata != nil {
-		metadata, err := json.Marshal(ute.metadata)
+	if uce.Name != "" {
+		val["name"] = uce.Name
+	}
+	if len(uce.Tags) > 0 {
+		tags := fmt.Sprintf("[%s]", strings.Join(uce.Tags, ","))
+		val["tags"] = tags
+	}
+	if uce.Credentials.Identity != "" {
+		val["identity"] = uce.Credentials.Identity
+	}
+	if uce.Metadata != nil {
+		metadata, err := json.Marshal(uce.Metadata)
 		if err != nil {
 			return val
 		}
 
-		val["metadata"] = string(metadata)
+		val["metadata"] = metadata
+	}
+	if uce.CreatedAt != (time.Time{}) {
+		val["created_at"] = uce.CreatedAt
+	}
+	if uce.Status.String() != "" {
+		val["status"] = uce.Status.String()
 	}
 
 	return val
 }
 
 type removeClientEvent struct {
-	id string
+	id        string
+	status    string
+	updatedAt time.Time
+	updatedBy string
+}
+
+func (rce removeClientEvent) Encode() map[string]interface{} {
+	return map[string]interface{}{
+		"operation":  clientRemove,
+		"id":         rce.id,
+		"status":     rce.status,
+		"updated_at": rce.updatedAt,
+		"updated_by": rce.updatedBy,
+	}
+}
+
+type shareClientEvent struct {
+	thingID string
+	actions string
+	userIDs string
+}
+
+func (sce shareClientEvent) Encode() map[string]interface{} {
+	return map[string]interface{}{
+		"operation": clientShare,
+		"thing_id":  sce.thingID,
+		"actions":   sce.actions,
+		"user_ids":  sce.userIDs,
+	}
+}
+
+type viewClientEvent struct {
+	mfclients.Client
+}
+
+func (vce viewClientEvent) Encode() map[string]interface{} {
```

Comment:
Would simple JSON encoding work here?

Reply:
I think we should do it that way.

Reply:
I have tried out different approached but none has yielded

```go
var val map[string]interface{}
data, err := json.Marshal(cce)
if err != nil {
	return val
}
if err := json.Unmarshal(data, &val); err != nil {
	return val
}
val["operation"] = clientCreate
return val
```

I get this error ` redis: can't marshal map[string]interface {} (implement encoding.BinaryMarshaler).`, even after adding `BinaryMarshaler` and `BinaryUnmarshaler` interfaces to clients i.e

```go
// MarshalBinary implements the encoding.BinaryMarshaler interface.
func (c *Client) MarshalBinary() ([]byte, error) {
	return json.Marshal(c)
}

// UnmarshalBinary implements the encoding.BinaryUnmarshaler interface.
func (c *Client) UnmarshalBinary(data []byte) error {
	return json.Unmarshal(data, c)
}
```